app/uk/gov/hmrc/gform/formtemplate/FormTemplateService.scala

```diff
@@ -86,19 +86,25 @@ class FormTemplateService(
 
     val exprSubstitutionsResolved: Either[UnexpectedState, ExprSubstitutions] = expressionsContext.resolveSelfReferences
 
+    def substitute(template: FormTemplate) =
+      for {
+        expressionsContextSubstituted <- fromOptA(exprSubstitutionsResolved)
+        substitutedFormTemplate = substituteExpressions(template, expressionsContextSubstituted)
+        substitutedBooleanExprsFormTemplate =
+          substituteBooleanExprs(
+            substitutedFormTemplate,
+            booleanExpressionsContext,
+            expressionsContextSubstituted
+          )
+        _                   <- verify(substitutedBooleanExprsFormTemplate)(expressionsContext)
+        formTemplateUpdated <- rewrite(substitutedBooleanExprsFormTemplate)
+      } yield formTemplateUpdated
+
     for {
-      expressionsContextSubstituted <- fromOptA(exprSubstitutionsResolved)
-      substitutedFormTemplate = substituteExpressions(formTemplate, expressionsContextSubstituted)
-      substitutedBooleanExprsFormTemplate =
-        substituteBooleanExprs(
-          substitutedFormTemplate,
-          booleanExpressionsContext,
-          expressionsContextSubstituted
-        )
-      _                  <- verify(substitutedBooleanExprsFormTemplate)(expressionsContext)
-      formTemplateToSave <- rewrite(substitutedBooleanExprsFormTemplate)
-      _                  <- formTemplateRepo.replace(mkSpecimen(formTemplateToSave))
-      _                  <- formRedirectRepo.deleteByFieldName("redirect", formTemplate._id.value)
+      formTemplateToSave         <- substitute(formTemplate)
+      formTemplateSpecimenToSave <- substitute(specimenSmartStrings(formTemplate))
```

Comment:
This seems to be missing functionality from last point in the description about top level expression being replaced by expression name